### PR TITLE
Add a new rule_type that allows to check the actions in a repo against a set of allowed actions

### PR DESCRIPTION
**Pull request by jhrozek**
Adds a new rule_type that allows specifying an allow list of actions
used in repositories. The evaluation uses rego and is able to print a
JSON representation of the actions that are not allowed.

Example profile:
```
version: v1
type: profile
name: actions-github-profile
context:
  provider: github
alert: "off"
remediate: "off"
repository:
  - type: repo_action_list
    def:
      actions:
        - actions/checkout
        - docker/build-push-action
        - docker/login-action
        - docker/metadata-action
        - docker/setup-buildx-action
        - sigstore/cosign-installer
```

Specifying an empty list allows you to list the actions used in the repo:
```
version: v1
type: profile
name: actions-github-profile
context:
  provider: github
alert: "off"
remediate: "off"
repository:
  - type: repo_action_list
    def:
      actions: []
```

Note that this PR depends on https://github.com/stacklok/minder/pull/1728

### CODE REVIEW

**Comment by jhrozek** [comment]
moved out of draft as the rule seems to work well

**Review by JAORMX — APPROVED** [review]

**Review by rdimitrov — APPROVED** [review]

File: rule-types/github/repo_action_list.yaml [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        package minder
+
+        expected_set := {x | x := input.profile.actions[_]}
+        actions := github_workflow.ls_actions("./.github/workflows")
+
+        violations[{"msg": msg}] {
+          intersection := expected_set & actions
+          not count(intersection) == count(actions)
+
+          difference := [x | x := actions[_]; not intersection[x]]
```
Comment (evankanderson) [inline rule-types/github/repo_action_list.yaml:46]: Can this simply be `expected_set - actions`?  If so, you could rewrite this as:

```suggestion
          missing := expected_set - actions
          not count(missing) == 0
          msg = format_message(missing, input.output_format)
```

**Review by evankanderson — APPROVED** [review]
I think I understand what this is doing, though rego branching feels a bit weird to me.

File: rule-types/github/repo_action_list.yaml (hunk 1 above)
Comment (jhrozek) [reply rule-types/github/repo_action_list.yaml:46]: I think we could use a combination of the two. The reason I used the weird list comprehension that creates the `difference` list was that IIRC I had issues formatting a set type later in the formatter. I think the difference you used reads better than the intersection, but I will also test edge cases (empty expected_set or expected_set that contains fewer items than actions). I'd like to use your proposal because it reads better, but I'm not that familiar with rego.

Comment (JAORMX) [reply rule-types/github/repo_action_list.yaml:46]: What if we merge the implementation as-is and push @evankanderson 's fix in a subsequent PR?

Comment (jhrozek) [reply rule-types/github/repo_action_list.yaml]: in my local testing a variant of what Evan suggested seems to work well. I'll give it a bit to test in staging

**Review by JAORMX — APPROVED** [review]
